### PR TITLE
Added i3 extractor and general dataconverter

### PR DESCRIPTION
**Pull request by RasmusOrsoe**

### CODE REVIEW

**Comment by asogaard** [comment]
Hi @RasmusOrsoe,
Just FYI: As this is a somewhat large PR (+788 lines), in the interest of merging the code relatively quickly, I will try to limit me feedback to overall code structure and functionality, and hold off on other stuff (linting, docstrings, unit tests, etc.). I am happy to work on this after this PR is merged.

File: src/data/sqlite_converter.py
Comment (asogaard) [inline src/data/sqlite_converter.py:85]: This seems like a somewhat involved function, and that might be necessary and if I please ignore this comment -- but at a first glance I would expect a combination of the `glob` and `re` packages and filtering to get most of the job done?

File: src/data/sqlite_converter.py
Comment (asogaard) [inline src/data/sqlite_converter.py:98]: I would consider giving this function a slightly more specific name; I would suggest the name I have suggested above for `walk_directory` so it might be worth considering if the two methods could be merged, provided `walk_directory` could be simplified sufficiently.

File: src/data/sqlite_converter.py
Comment (asogaard) [inline src/data/sqlite_converter.py:124]: The return value of this function does not seem to be used anywhere. Given that, is this function strictly necessary or could it be replaced by calls to `os.makedirs(outdir, exists_ok=True)`? Otherwise, perhaps drop the return values from this function and consider producing a slightly less dire warning. ;)

File: src/data/sqlite_converter.py
Comment (asogaard) [inline src/data/sqlite_converter.py:140]: From here and down to L323 seems to be functionality that is specific to the `SQLiteDataConverter` class; if so I would recommend moving these to (private) class methods.

File: src/data/sqlite_converter.py
Comment (asogaard) [inline src/data/sqlite_converter.py:337]: Right now this class does not use the `process_file` function of the base class (NB: I realise that this class does not, indeed, inherit from `DataConverter` right now, but eventually it will) and it does not seem to implement the `_save` method. This means that it deviates from the syntax "promised" by the base class. Therefore, I think that we should discuss whether the proposed syntax in `DataConverter` makes sense from an implementation standpoint; if so, we should rewrite the `SQLiteDataConverter` class to conform to this; if not, we should rewrite the base class to be a better fit to the actual implementation(s). Let me know what you think.

File: src/data/test_sqlite.py
Comment (asogaard) [inline src/data/test_sqlite.py:1]: Fantastic with a test script! I would move it to a separate `test` directory, write it as a unit test, and package it with minimal data files required for testing -- but let's leave that till after this PR.

**Review by asogaard — CHANGES_REQUESTED** [review]
Hi @RasmusOrsoe,
Thanks for adding all of this hard work to our common repo! 🥳   Brilliant to get the ball rolling like this, and I look forward to getting the first batch of material into `main`.

I have tried to proved some comments and suggestions for each of the four added files. Please let me know if anything is unclear, if you disagree with my suggestions, and if there is anything I can do to help out.

File: src/data/dataconverter.py
Comment (RasmusOrsoe) [reply src/data/dataconverter.py:6]: Changed according to your recommendation.

File: src/data/dataconverter.py
Comment (RasmusOrsoe) [reply src/data/dataconverter.py:9]: All inputs now flow to the I3Extractor

File: src/data/dataconverter.py
Comment (RasmusOrsoe) [reply src/data/dataconverter.py:26]: This class is indeed in a non-usable state. 

I had trouble getting the functionality I wanted from this general class for the i3dataconverter, so I threw away the inheritance of the general class and started building a class that has the functionality I require. I believe it will be easier to define a general class from which we can inherit from when we have one that actually works, which we do now. 

One of the issues I had was that it is restricting for the general class to require a method that process an entire file. I'd like a method that works on a single frame instead, and that it should be class-specific on how we choose to process the files. All we need to inherit is the way in which each frame is read.


.. just my thoughts.

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:1]: renamed

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:17]: I doubt these are usefull outside the scope of this class.

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:23]: renamed

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:40]: renamed

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:48]: renamed

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:85]: Maybe. 

Lets zoom about this some time

Rasmus

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:98]: Leaving this for a rainy day

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:124]: Return value is used at least once. But the check is not strictly neccesary. 

The method you mention is not compatible with python >3.0 and by using error handling we get the same functionality.

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:140]: Two distinct functionalities are in those lines: 

1) The creation of temporary .db-files which are done in parallel
2) The merging of said temporary databases files into a single, final database.

I cannot for the love of god put a private function (e.g. a function defined within the class) as an argument for map_asyinc

```
if __name__ == 'sqlite_converter':
              #parallel_extraction(settings[0])
              print('starting pool!')
              p = Pool(processes = workers)
              p.map_async(parallel_extraction, settings)
              p.close()
              p.join()
              print('pool')
              self._merge_databases()
```

When I tried to make `parallel_extraction` a private function, the script class would act as if it simply never called the function. However, this isn't whats going on; the function call for every worker fails (without error messages thx to multiprocessing). From googling I found that this is likely due to `self` not being handled correctly. If we were to obtain the error message, I'd think it would be something like:  `parallel_extraction recieved two arguments but expected 1`.

I **cannot** fix this. If we want it fixed someone needs to come up with a good idea.

Rasmus

File: src/data/sqlite_converter.py
Comment (RasmusOrsoe) [reply src/data/sqlite_converter.py:337]: Yes I'm breaking the rules already!

I think we should take a look at this class and from it decipher where we can generalize syntax for a practical general dataconverter class.

Rasmus

File: src/data/test_sqlite.py
Comment (RasmusOrsoe) [reply src/data/test_sqlite.py:1]: Agreed. I did it this way because you cannot import from .py-files that are not in the same folder as the test scripts - unless it is installed as a package, which I did not want to mess with right now.


Rasmus

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:37]: Added get_om_keys()

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:89]: Good to know! 

I did indeed extract these quantities like extract_features in the past, but my experience is that it is a quite annyoing format for when you want to change the extracted quantities. Also, the dictionary field names ends up being the column names in the sqlite database. So for swiftness, a user could pass something like the above and it would work flawlessly.

Right now, when an evaluation fails, it simply pads the failed value with default -1.

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:128]: renamed and moved into i3extractor.

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:144]: saving this for a rainy day.

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:175]: moved

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:193]: moved

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:254]: moved to i3extractor

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:290]: renamed and moved to i3extractor

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:298]: moved to i3extractor

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:305]: moved

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:333]: I dont know how to shorten this.

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:346]: done

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:272]: moved

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:3]: moved

File: src/data/utils.py
Comment (RasmusOrsoe) [reply src/data/utils.py:1]: moved extractor specific functions into i3extractor class.

**Comment by RasmusOrsoe** [comment]
I have addressed the vast majority of issues and I vote that we for now live with the fact that the parallelized function in the i3dataconverter lives outside the class.

File: src/data/sqlite_converter.py
Comment (asogaard) [reply src/data/sqlite_converter.py]: No worries, and thanks for the detailed description. I have created #11 and will try to work on this problem there.

**Comment by asogaard** [comment]
Hi @RasmusOrsoe,
Thanks for all of the comments and the code changes! I complete agree that we merge this PR and follow up on whatever remains through separate PRs/issues. Great job! 🥳

**Review by asogaard — APPROVED** [review]

**Pull request merged**
